src/amo/experiments/20210531_download_funnel_experiment.js

```diff
@@ -1,8 +1,11 @@
 /* @flow */
+import { FIREFOX_USERS } from 'amo/constants';
+
 export const VARIANT_CURRENT = 'current-link';
 export const VARIANT_NEW = 'new-link';
 
 export const EXPERIMENT_CONFIG = {
+  excludedGroups: [FIREFOX_USERS],
```

Comment:
I think an experiment config should expose a function instead of this array. It would take the state as input and return a boolean. We could have some helper functions for checks we think we'll reuse in the future. Something along these lines:

```js
export const EXPERIMENT_CONFIG = {
  id: '20210531_download_funnel_experiment',
  variants: [
	{ id: VARIANT_CURRENT, percentage: 0.5 },
  	{ id: VARIANT_NEW, percentage: 0.5 },
  ],
  shouldNotBeInExperiment({ state }) {
    const { userAgentInfo } = state;
    
    return _isFirefox({ userAgentInfo });
  },
};
```

This is easy to follow and adding more conditions to that experiment shouldn't be a problem compared to the approach with new constants, an array and logic to match each constant with each user-specific data. Unless I am missing something, this experiment config describes precisely the experiment, which is what we want.

I'd use this function in the `mapStateToProps()` function of the HOC:

```js
  const mapStateToProps = (
      state: AppState,
    ): WithExperimentsPropsFromState => {
      return {
        storedVariants: state.experiments,
		shouldNotBeInExperiment: shouldNotBeInExperiment({ state }),
      };
    };
```

From there, I think we can write:

```js
// Otherwise if the user is to be excluded, use the NOT_IN_EXPERIMENT variant.
else if (shouldNotBeInExperiment) {
  variant = NOT_IN_EXPERIMENT;
}
```

Reply:
Tell me if I am wrong but I think implementing https://github.com/mozilla/addons-frontend/issues/10686 would then be:

```js
  shouldNotBeInExperiment({ state }) {
    const { userAgentInfo, api } = state;
    
    return api.lang !== 'en-US' || !_isFirefox({ userAgentInfo });
  },
```

Reply:
Thanks for the suggestion @willdurand. I had considered something similar, where each experiment would define a function, and then we would have unlimited flexibility. The reason I went the route I did was:

1. From the perspective of this being a framework, I was hoping to restrict the experiment config files to just being config, and not having any executable code. This would make implementing future experiments quicker and easier.
2. Similar to the above, I was thinking we'd take the "development hit" now, deal with all of the complexity in the framework, and although it's more code now, it means less code to write and less to think about when creating new experiments in the future. It also means all of the tests for the logic could be written now and we wouldn't need new tests for new experiments.

I do, however, like your implementation where the `shouldNotBeInExperiment` function accepts a `state` argument, which allows for a lot of future flexibility, and I suppose we are not building a framework that we expect to be used by people who need it to be simple - we're building the framework for ourselves to use, and I think we can handle defining a function in the experiment config file. Therefore I think your suggestion is better overall and I will make changes to implement it.

Would you imaging landing a unit test along with each new experiment that would test the `shouldNotBeInExperiment` that is defined in the config file, or do you think that's something we would simply test manually and expect QA to verify?

Reply:
> Would you imaging landing a unit test along with each new experiment that would test the `shouldNotBeInExperiment` that is defined in the config file, or do you think that's something we would simply test manually and expect QA to verify?

I'd write a test case (or two) for the experiment config to make sure the exclusion logic is correct.